### ISSUE TITLE
Numeric diff of BatchNormalization layer when migrate from Keras2 to keras3

### ISSUE THREAD

**Issue by cliu-nuro** [keras-team/keras#19649]
Hi, I'm noticing that `BatchNormalization` layer generates different output when migrated from Keras2 to Keras3. Is this expected?

To reproduce,
1. Run the following script with keras2 with `--generate`
2. Run the script again with keras2 without `--generate`, to see that the result is deterministic
3. Run the script again with keras3 without `--generate`, see that as the input gets larger, we start to see discrepancy.

I played around with `scale` or training params, neither of them seems to be directly related. It feels like as more computation gets involves (e.g. changing `scale` from `False` to `True`), it will become easier to expose the discrepancy.
```
import os
import random

import gflags
from packaging.version import Version
import numpy as np

import keras
if Version(keras.__version__).major == 2:
    from tensorflow import keras
import tensorflow as tf

gflags.DEFINE_bool("generate", False, "Whether to generate instead of check.")

FLAGS = gflags.FLAGS

def save_or_compare_bn_output(
        generate_mode: bool,
        use_small_image: bool = False,
        training: bool = False,
):
    """Construct a deterministic input and pass it to the BatchNormalization
    layer. Save or compare the output based on `generate_mode`.
    """
    if use_small_image:
        shape = (16, 24, 24)
    else:
        shape = (16, 302, 480)
    desired_mean = tf.constant([
        8.4438013e+02,  8.0877607e+03,  8.2639297e+01,  4.7700565e+02,
        1.1063718e+02, -4.4651328e+03, -6.6082535e+00,  1.6212619e+02,
        5.7152104e+03, -7.6639697e+02, -6.9809601e+01,  8.2015007e+01,
        6.4822838e+01,  1.5333386e+02, -2.1635346e+00, -1.1024360e+04
    ], dtype=tf.float32)
    desired_mean = tf.reshape(desired_mean, [16, 1, 1])

    desired_variance = tf.constant([
        1.3268768e+06, 6.9939104e+07, 2.7830322e+06, 1.7201350e+07,
       5.2418816e+05, 2.3168248e+07, 2.0690480e+04, 3.8199915e+06,
       5.8451648e+07, 9.0565344e+05, 2.3908058e+06, 3.9144145e+06,
       3.4122118e+06, 3.5702670e+06, 2.2716367e+04, 1.2985819e+08
    ], dtype=tf.float32)
    desired_stddev = tf.sqrt(desired_variance)
    desired_stddev = tf.reshape(desired_stddev, [16, 1, 1])
    random_tensor = tf.random.normal(
        shape=shape, mean=desired_mean, stddev=desired_stddev)
    deterministic_input = tf.expand_dims(random_tensor, axis=0)

    gamma_init = tf.constant_initializer([
        0.5737993 , 1.7945665 , 0.31027183, 1.6139743 , 0.4056443 ,
        1.2973828 , 0.3193084 , 0.31500784, 1.8060954 , 0.39719   ,
        0.24610205, 0.38457012, 0.39991254, 0.34851488, 0.30739728,
        2.2470322
    ])
    beta_init = tf.constant_initializer([
        1.2459143e+00, -1.5146567e+00,  5.9464660e-02, -6.6626422e-02,
        6.6803358e-01, -1.3482059e+00, -2.5516704e-03, -3.7482270e-04,
        1.8988891e+00,  2.9501798e+00,  2.7108081e-02, -6.6227927e-03,
        9.3242424e-03,  4.5773465e-02,  5.7123094e-03,  1.8978829e+00
    ])
    moving_mean_init = tf.constant_initializer([
        8.4438013e+02,  8.0877607e+03,  8.2639297e+01,  4.7700565e+02,
        1.1063718e+02, -4.4651328e+03, -6.6082535e+00,  1.6212619e+02,
        7.7152104e+03, -7.6639697e+02, -6.9809601e+01,  8.2015007e+01,
        8.4822838e+01,  1.5333386e+02, -2.1635346e+00, -1.1024360e+04
    ])
    moving_variance_init = tf.constant_initializer([
        1.3268768e+06, 6.9939104e+07, 2.7830322e+06, 1.7201350e+07,
        5.2418816e+05, 2.3168248e+07, 2.0690480e+04, 3.8199915e+06,
        4.8451648e+07, 9.0565344e+05, 2.3908058e+06, 3.9144145e+06,
        2.4122118e+06, 3.5702670e+06, 2.2716367e+04, 1.2985819e+08
    ])
    bn_layer = keras.layers.BatchNormalization(
        axis=1,
        trainable=training,
        gamma_initializer=gamma_init,
        beta_initializer=beta_init,
        moving_mean_initializer=moving_mean_init,
        moving_variance_initializer=moving_variance_init,
    )
    bn_output = bn_layer(deterministic_input, training=training)

    prefix = 'small_input' if use_small_image else 'large_input'
    prefix += '_train' if training else '_predict'
    name = f'{prefix}_bn_output'
    filename = f'/tmp/keras_bn/{name}.blob'
    if generate_mode:
        _dump_float_tensor_to_blob_file(bn_output.numpy(), filename)
    else:
        test_tensor = bn_output.numpy()
        golden_tensor = _get_float_tensor_from_blob_file(filename)
        _compare_tensor(test_tensor, golden_tensor)

def _dump_float_tensor_to_blob_file(data: np.ndarray, filepath: str):
    data_size = np.prod(data.shape)
    data = np.concatenate([[len(data.shape), data_size], data.shape,
                           np.reshape(data, (-1))])
    data = data.astype(np.float32)
    data.tofile(filepath)

def _get_float_tensor_from_blob_file(filepath: str) -> np.ndarray:
    data = np.fromfile(filepath, dtype=np.float32)
    dim = int(data[0])
    count = int(data[1])
    shape = data[2:2 + dim].astype(int)
    return data[2 + dim:].reshape(shape)

def _compare_tensor(test_tensor: np.ndarray, golden_tensor: np.ndarray):
    np.testing.assert_allclose(test_tensor,
                               golden_tensor,
                               err_msg='BN output mismatch',
                               verbose=True,
                               atol=1e-7,
                               rtol=1e-3,
                               equal_nan=True)

def main(argv):
    assert len(argv) == 1, argv
    for use_small_image in [True, False]:
        for training in [False, True]:
            try:
                save_or_compare_bn_output(
                    generate_mode=FLAGS.generate,
                    use_small_image=use_small_image,
                    training=training,
                )
                if not FLAGS.generate:
                    print(f'PASS ===============> for small input: {use_small_image}, training: {training}')
            except AssertionError:
                if not FLAGS.generate:
                    print(f'FAIL ===============> for small input: {use_small_image}, training: {training}')



if __name__ == "__main__":
    os.environ['TF_CUDNN_DETERMINISTIC'] = '1'
    os.environ['TF_DETERMINISTIC_OPS'] = '1'

    # Fix rand seeds.
    random.seed(12345)
    # np random doesn't seem to be necessary.
    # np.random.seed(12345)
    tf.random.set_seed(12345)

    # Change this if you want to use a specific GPU.
    # os.environ['CUDA_VISIBLE_DEVICES'] = '0'
    main()
```

**Comment by fchollet** [comment]
This is a very complex script -- here's a simpler one to check Keras 2 vs Keras 3 BN numerics:

```python
import numpy as np
import keras
import tf_keras

keras2_bn = tf_keras.layers.BatchNormalization()
keras3_bn = keras.layers.BatchNormalization()

keras2_bn.build((None, 128, 128, 3))
keras3_bn.build((None, 128, 128, 3))

keras2_bn.beta.assign(np.random.random(keras2_bn.beta.shape))
keras2_bn.gamma.assign(np.random.random(keras2_bn.gamma.shape))
keras2_bn.moving_mean.assign(np.random.random(keras2_bn.moving_mean.shape))
keras2_bn.moving_variance.assign(np.random.random(keras2_bn.moving_variance.shape))
keras3_bn.set_weights(keras2_bn.get_weights())

x = np.random.random((32, 128, 128, 3))

out_2_t = keras2_bn(x, training=True)
out_3_t = keras3_bn(x, training=True)

out_2 = keras2_bn(x)
out_3 = keras3_bn(x)

print(np.max(out_2_t - out_3_t))
print(np.max(out_2 - out_3))
```

This script indicates that the maximum per-coefficient divergence between the two is around 1e-7 in float32, and 1e-11 in float64, which is what's expected. float32 computation is never more precise than that.

**Comment by haohuanw** [comment]
@fchollet the main thoughts are the output should ideally be identical instead of having minor differences.

i took a further look and found the caveat. keras2 will use fused bn whenever possible and calling `tf.nn.fused_batch_norm` and keras3 always call `tf.nn.batch_normalization`.

setting `keras2_bn = tf_keras.layers.BatchNormalization(fused=False)` will produce identical results between keras2 & 3.

i don't think this is an issue and the difference is acceptable. we will workaround this when we are testing parity between 2 & 3.

**Comment by fchollet** [comment]
> the main thoughts are the output should ideally be identical instead of having minor differences.

This is impossible unless they are exactly the same kernels run in the same order -- but Keras 3 is a different framework, that runs on multiple backends, so it cannot work like this.

If you take two mathematically-equivalent expressions that are factored slightly differently (e.g. different operation order...), they will only match up to 1e-7 when run in float32 (except in special cases). Despite being mathematically equivalent. that is simply how float32 computation works. It isn't infinitely precise.

Likewise Keras 3 layers only match Keras 2 layers up to 1e-7, and further, Keras 3 layers only match themselves when run in a different backend up to 1e-7.

**Comment by haohuanw** [comment]
> This is impossible unless they are exactly the same kernels run in the same order

yes - the missing part is that we had always thought that for tf backend on BN it uses same kernel but apparently is not.

other statements on match guarantees make sense.

**Comment by cliu-nuro** [comment]
> they will only match up to 1e-7 when run in float32 (except in special cases). Despite being mathematically equivalent.

@fchollet Thanks for the clarification. 

The precision you mentioned above is exactly why my script is so complicated. `atol=1e-7` and `rtol=1e-3` are what I used for float32. But I'm seeing much larger difference `atol=1e-5` and `rtol=1e-1` for a layer I was testing.
When I tried to reproduce it with some simple example, I figured that I cannot reproduce it. So I tried to mimic the real values I saw in my test.

I might not go further on this given the kernels will be different and we don't expect the same result.
But I want to flag here that `match up to 1e-7` doesn't always hold. It heavily depends on the input.
